### PR TITLE
get rid of pynsist, create Porcupine.exe and registry entries for opening with porcupine

### LINKED ISSUE #637: Associate python files (and possibly other codes) with porcupine
Arrinao tried to do this and gave up

```
<Arrinao> where's the freaking porcupine executable??
<Arrinao> there's none over here
<Arrinao> C:\Users\Martin\AppData\Local\Programs\Porcupine
<Akuli> why do you need to know?
<Arrinao> I want to set up a registry entry to make porcupine open .py files
<Akuli> you can right-click porcupine in start menu
<Akuli> that's a good idea though, please create an issue about it :)
<Akuli> maybe porcupine should do it by default in windows
<Akuli> if you right-click and choose properties, it should show
<Arrinao> I can, but that doesn't point me at an executable but at a shortcut
<Akuli> so where does the shortcut point? look at its properties
<Arrinao> it points into the directory I've just posted :D
<Arrinao> but there's no executable
<Arrinao> C:\Users\Martin\AppData\Local\Programs\Porcupine\Python\pythonw.exe "C:\Users\Martin\AppData\Local\Programs\Porcupine\Porcupine.launch.pyw"
<Akuli> so... did you read the path?
<Arrinao> I've read only the first part, it didn' display as a whole until I copy pasted here
<Akuli> lol
<Arrinao> phew
<Arrinao> so what now?
<Akuli> to open a file, you need to call ...\Porcupine\Python\pythonw.exe with two arguments. ...\Porcupine\Porcupine.launch.pyw and the file you want to edit
<Akuli> i have no idea how you teach that to windows registry :)
<Arrinao> I don't even think I can
<Arrinao> whatever
<Akuli> i'll make the issue for you
```

### LINKED ISSUE #650: Porcupine shows up as pythonw.exe in task manager
Would be better as porcupine.exe. I don't think I can just rename it, because it seems that the exe file must have Porcupine's icon and description for #637

### PR DESCRIPTION
**Pull request by Akuli**
Fixes #650 
Helps with #637  but needs more registry keys (can compare to what firefox does)

- [ ] error is msvcrt not installed (needed for python)
- [x] make porcupine icon show up in uninstall list again (used to work, something broke)
- [x] figure out why it shows up as "Python" when trying to open files with it
- [x] make subprocesses work
- [x] figure out why "Destination folder" became blank by default (used to work)

### CODE REVIEW

File: scripts/installer.nsi [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+!insertmacro MUI_PAGE_WELCOME
+!insertmacro MUI_PAGE_LICENSE LICENSE
+!insertmacro MULTIUSER_PAGE_INSTALLMODE
+!insertmacro MUI_PAGE_DIRECTORY
+!insertmacro MUI_PAGE_INSTFILES
+!insertmacro MUI_PAGE_FINISH
+!insertmacro MUI_LANGUAGE "English"
+
+Name "Porcupine ${VERSION}"
+OutFile "Porcupine_${VERSION}.exe"
```
Comment (Akuli) [inline scripts/installer.nsi:40]: Maybe rename this to `PorcupineSetup_123.456.789.exe`?

File: scripts/installer.nsi [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+Section "!Porcupine" sec_app
+  SetRegView 64
+  SectionIn RO
+
+  ; Marker file for per-user install
+  StrCmp $MultiUser.InstallMode CurrentUser 0 +3
+    FileOpen $0 "$INSTDIR\${USER_INSTALL_MARKER}" w
+    FileClose $0
+
+  ; Install files
```
Comment (Akuli) [inline scripts/installer.nsi:60]:
```suggestion
```

File: scripts/installer.nsi [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+  File /r "pkgs\*.*"
+  SetOutPath "$INSTDIR\Python"
+  File /r "Python\*.*"
+  SetOutPath "$INSTDIR\lib"
+  File /r "lib\*.*"
+  SetOutPath "$INSTDIR"
+
+  DetailPrint "Creating shortcut..."
+  SetOutPath "%HOMEDRIVE%\%HOMEPATH%"  ; This becomes working directory for shortcut
+  CreateShortCut "$SMPROGRAMS\Porcupine.lnk" '"$INSTDIR\Python\Porcupine.exe"' "$INSTDIR\Python\Porcupine.exe"
```
Comment (Akuli) [inline scripts/installer.nsi:72]:
```suggestion
  SetOutPath "%HOMEDRIVE%\%HOMEPATH%"  ; This becomes working directory for shortcut
  CreateShortCut "$SMPROGRAMS\Porcupine.lnk" '"$INSTDIR\Python\Porcupine.exe"' "$INSTDIR\Python\Porcupine.exe"
  SetOutPath "$INSTDIR"
```

File: scripts/installer.nsi [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+  IfFileExists "$INSTDIR\${USER_INSTALL_MARKER}" 0 +3
+    SetShellVarContext current
+    Delete "$INSTDIR\${USER_INSTALL_MARKER}"
+
+  Delete "$INSTDIR\uninstall.exe"
+  Delete "$INSTDIR\launch.pyw"
+  RMDir /r "$INSTDIR\pkgs"
+  RMDir /r "$INSTDIR\Python"
+  RMDir /r "$INSTDIR\lib"
+  RMDir "$INSTDIR"
```
Comment (Akuli) [inline scripts/installer.nsi:103]:
```suggestion
  RMDir /r "$INSTDIR"
```

File: scripts/installer.nsi [hunk 5]
```diff
@@ -0,0 +1,147 @@
+; Based on a file that pynsist generated
+; Definition for VERSION must be added before running this file
+
+!define BITNESS "64"
+!define ARCH_TAG ".amd64"
+
+; Marker file to tell the uninstaller that it's a user installation
+!define USER_INSTALL_MARKER _user_install_marker
```
Comment (Akuli) [inline scripts/installer.nsi:8]: This variable seems a bit unnecessary. Might as well inline it

File: scripts/installer.nsi [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  RMDir /r "$INSTDIR\lib"
+  RMDir "$INSTDIR"
+
+  Delete "$SMPROGRAMS\Porcupine.lnk"
+  DeleteRegKey SHCTX "Software\Microsoft\Windows\CurrentVersion\Uninstall\Porcupine"
+  DeleteRegKey SHCTX "Software\Classes\Applications\Porcupine.exe"
+SectionEnd
+
+
+Function .onMouseOverSection
```
Comment (Akuli) [inline scripts/installer.nsi:111]: Can this section be deleted? If not:

```suggestion
; I have no idea what this function does. It was created by nsist.
Function .onMouseOverSection
```

File: scripts/installer.nsi [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+SetCompressor lzma
+
+Unicode true
+ManifestDPIAware true
+
+!define MULTIUSER_EXECUTIONLEVEL Highest
+!define MULTIUSER_INSTALLMODE_DEFAULT_CURRENTUSER
+!define MULTIUSER_MUI
+!define MULTIUSER_INSTALLMODE_COMMANDLINE
+!define MULTIUSER_INSTALLMODE_INSTDIR "Porcupine"
```
Comment (Akuli) [inline scripts/installer.nsi:19]: This variable is a bit useless, should inline

**Comment by Akuli** [comment]
I have now tried several different options:
- `setup-mingw` action: works on github actions but not usable locally
- `zig cc` does not work on windows 7 (needed LLVM installed separately just to get resource compiler)
- `scoop install gcc`: https://github.com/lukesampson/scoop/issues/4235
- The MinGW installer gui thingy: appears to be gui only, does not work even locally
- MSVC: not going to go closed-source if I don't really really have to

Tl;dr I just can't get a C compiler to work on windows, although I can use `setup-mingw` on github actions.

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -2,25 +2,58 @@ on:
...
 jobs:
-  windows_build:
+  windows-build:
     runs-on: windows-latest
     steps:
     - uses: actions/checkout@v2
     - uses: actions/setup-python@v1
       with:
-        python-version: ${{ matrix.python-version }}
+        python-version: "3.9"
```
Comment (Akuli) [inline .github/workflows/build.yml:14]:
```suggestion
        python-version: "3.8"  # last version supporting windows 7
```

**Comment by Akuli** [comment]
The "Open with" dialog has been fucking annoying for the past day or so. Turns out that it has a cache in registry, in `HKEY_CURRENT_USER\Software\Classes\Local Settings\Software\Microsoft\Windows\Shell\MuiCache`. I wish I had known this earlier.

File: scripts/installer.nsi (hunk 6 above)
Comment (Akuli) [reply scripts/installer.nsi:101]: deleted

File: docs/utils.rst [hunk 9]
```diff
@@ -16,7 +16,9 @@ Information about Python
 
     The ``pythonw.exe`` program runs Python scripts without a command
     prompt, so you need to check for that when doing things like
-    starting a new command prompt from Python.
+    starting a new command prompt from Python. This is also ``True``
+    if Porcupine is running as ``Porcupine.exe``, which it is when
+    launched when launched from start menu.
```
Comment (Akuli) [inline docs/utils.rst:21]:
```suggestion
    launched from start menu.
```

File: launcher/launch.pyw [hunk 10]
```diff
@@ -0,0 +1,3 @@
+from porcupine.__main__ import main
```
Comment (Akuli) [inline launcher/launch.pyw:1]:
```suggestion
# Can't use -m or -c in main.c, would import from current working directory (user's home)
from porcupine.__main__ import main
```

File: launcher/main.c [hunk 11]
```diff
@@ -0,0 +1,47 @@
+#include <assert.h>
```
Comment (Akuli) [inline launcher/main.c:1]:
```suggestion
```

File: porcupine/default_filetypes.toml [hunk 12]
```diff
@@ -315,3 +315,10 @@ filename_patterns = ["*.tex"]
 pygments_lexer = "pygments.lexers.TexLexer"
 tabs2spaces = true
 comment_prefix = '%'
+
+["NSIS installer file"]
+filename_patterns = ["*.nsi", "*.nsh"]
+pygments_lexer = "pygments.lexers.NSISLexer"
+tabs2spaces = true
+indent_size = 2
+comment_prefix = '%'
```
Comment (Akuli) [inline porcupine/default_filetypes.toml:324]:
```suggestion
comment_prefix = ';'
```

File: porcupine/plugins/poppingtabs.py [hunk 13]
```diff
@@ -134,7 +134,7 @@ def pop(self, tab: tabs.Tab, state: Any, geometry: str) -> None:
         # e.g. queue.py (#31), but good when that's where porcupine is
         # meant to be imported from (#230).
         code = f"import sys; sys.path[:] = {sys.path}; from porcupine.__main__ import main; main()"
-        args = [sys.executable, "-c", code]
+        args = [sys.executable, "-c", code]  # FIXME: sys.executable.endswith(r"\Porcupine.exe")
```
Comment (Akuli) [inline porcupine/plugins/poppingtabs.py:137]: Should create issue

File: scripts/build-exe-installer.py [hunk 14]
```diff
@@ -1,129 +1,135 @@
...
 
-    parser.write(sys.stderr)
-    sys.stderr.flush()
+# https://pynsist.readthedocs.io/en/latest/faq.html#packaging-with-tkinter
+# We don't use pynsist because it does not allow specifying a custom executable.
+# We have a custom Porcupine.exe launcher which has the custom icon and can be called
+# with no arguments to launch Porcupine.
+# I couldn't get python to import from anywhere else than from Python directory, no separate pynsist_pkgs
+shutil.copytree(prefix / "tcl", "build/lib")
+for file in list((prefix / "DLLs").glob("tk*.dll")) + list((prefix / "DLLs").glob("tcl*.dll")):
```
Comment (Akuli) [inline scripts/build-exe-installer.py:60]: new-ish syntax

```suggestion
for file in [*(prefix / "DLLs").glob("tk*.dll"), *(prefix / "DLLs").glob("tcl*.dll")]:
```

File: scripts/build-exe-installer.py [hunk 15]
```diff
@@ -1,129 +1,135 @@
...
+    shutil.copy(file, "build/Python")
+shutil.copy(prefix / "DLLs" / "_tkinter.pyd", "build/Python")
+shutil.copy(prefix / "libs" / "_tkinter.lib", "build/Python")
+shutil.copytree(tkinter.__path__[0], "build/Python/tkinter")
+
+shutil.copy("scripts/installer.nsi", "build/installer.nsi")
+shutil.copy("LICENSE", "build/LICENSE")
+shutil.copytree("launcher", "build/launcher")
+
+metadata_file = Path("build/launcher/metadata.rc")
```
Comment (Akuli) [inline scripts/build-exe-installer.py:70]:
```suggestion
# I tried to give -D option to preprocessor, didn't work
metadata_file = Path("build/launcher/metadata.rc")
```

File: scripts/build-exe-installer.py [hunk 16]
```diff
@@ -1,129 +1,135 @@
...
+            "Porcupine.exe",
+            "main.c",
+            "icon.res",
+            "metadata.res",
+        ],
+        cwd="build/launcher",
+    )
 
+print("Installing Porcupine into build/Python with pip")
+# pyls needs pkg_resources which comes with setuptools, don't know why it doesn't install by default
```
Comment (Akuli) [inline scripts/build-exe-installer.py:110]:
```suggestion
# pyls needs pkg_resources from setuptools, don't know why it doesn't install by default
```

File: scripts/installer.nsi [hunk 17]
```diff
@@ -0,0 +1,103 @@
+; Based on a file that pynsist generated
+
+!define BITNESS "64"
+!define ARCH_TAG ".amd64"
```
Comment (Akuli) [inline scripts/installer.nsi:4]: Are these necessary?

File: scripts/installer.nsi [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+  WriteRegDWORD SHCTX "Software\Microsoft\Windows\CurrentVersion\Uninstall\Porcupine" "NoModify" 1
+  WriteRegDWORD SHCTX "Software\Microsoft\Windows\CurrentVersion\Uninstall\Porcupine" "NoRepair" 1
+SectionEnd
+
+Section "Uninstall"
+  SetRegView 64
+  SetShellVarContext all
+  IfFileExists "$INSTDIR\_user_install_marker" 0 +3
+    SetShellVarContext current
+    Delete "$INSTDIR\_user_install_marker"
```
Comment (Akuli) [inline scripts/installer.nsi:88]: The whole install dir will be deleted anyway:

```suggestion
  IfFileExists "$INSTDIR\_user_install_marker" 0 +2
    SetShellVarContext current
```

File: porcupine/plugins/poppingtabs.py (hunk 13 above)
Comment (Akuli) [reply porcupine/plugins/poppingtabs.py:137]: #654

File: scripts/installer.nsi (hunk 17 above)
Comment (Akuli) [reply scripts/installer.nsi]: no
